tower-balance/src/lib.rs

```diff
@@ -1,337 +1,287 @@
-    /// dispatch the next request.
-    chosen_ready_index: Option<usize>,
+pub struct P2CBalance<D: Discover> {
+    // XXX Pool requires direct access to this... Not ideal.
+    pub(crate) discover: D,
 
-    /// Holds an index into `ready`, indicating the service that dispatched the last
-    /// request.
-    dispatched_ready_index: Option<usize>,
+    endpoints: IndexMap<D::Key, D::Service>,
 
-    /// Holds all possibly-available endpoints (i.e. from `discover`).
-    ready: IndexMap<D::Key, D::Service>,
+    /// Holds an index into `endpoints`, indicating the service that has been
+    /// chosen to dispatch the next request.
+    ready_index: Option<usize>,
 
-    /// Newly-added endpoints that have not yet become ready.
-    not_ready: IndexMap<D::Key, D::Service>,
+    rng: SmallRng,
 }
 
-// ===== impl Balance =====
-
-impl<D> Balance<D, choose::PowerOfTwoChoices>
-where
-    D: Discover,
-    D::Service: Load,
-    <D::Service as Load>::Metric: PartialOrd + fmt::Debug,
-{
-    /// Chooses services using the [Power of Two Choices][p2c].
-    ///
-    /// This configuration is prefered when a load metric is known.
-    ///
-    /// As described in the [Finagle Guide][finagle]:
-    ///
-    /// > The algorithm randomly picks two services from the set of ready endpoints and
-    /// > selects the least loaded of the two. By repeatedly using this strategy, we can
-    /// > expect a manageable upper bound on the maximum load of any server.
-    /// >
-    /// > The maximum load variance between any two servers is bound by `ln(ln(n))` where
-    /// > `n` is the number of servers in the cluster.
-    ///
-    /// [finagle]: https://twitter.github.io/finagle/guide/Clients.html#power-of-two-choices-p2c-least-loaded
-    /// [p2c]: http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
-    pub fn p2c(discover: D) -> Self {
-        Self::new(discover, choose::PowerOfTwoChoices::default())
-    }
+// ===== impl P2CBalance =====
 
+impl<D: Discover> P2CBalance<D> {
     /// Initializes a P2C load balancer from the provided randomization source.
-    ///
-    /// This may be preferable when an application instantiates many balancers.
-    pub fn p2c_with_rng<R: rand::Rng>(discover: D, rng: &mut R) -> Result<Self, rand::Error> {
-        let rng = SmallRng::from_rng(rng)?;
-        Ok(Self::new(discover, choose::PowerOfTwoChoices::new(rng)))
-    }
-}
-
-impl<D: Discover> Balance<D, choose::RoundRobin> {
-    /// Attempts to choose services sequentially.
-    ///
-    /// This configuration is prefered when no load metric is known.
-    pub fn round_robin(discover: D) -> Self {
-        Self::new(discover, choose::RoundRobin::default())
-    }
-}
-
-impl<D, C> Balance<D, C>
-where
-    D: Discover,
-    C: Choose<D::Key, D::Service>,
-{
-    /// Creates a new balancer.
-    pub fn new(discover: D, choose: C) -> Self {
+    pub fn new(discover: D, rng: SmallRng) -> Self {
```

Comment:
This is quite nitpicky, but I'm not sure whether it's more idiomatic for this to be `new`, or to make this `new_with_rng` and call `from_entropy` `new`?

Reply:
This isn't meant to be primary entry-point (the layer is), so i ended up with this. In the proxy, for instance, we weren't reusing the RNGs and were creating a new seed for each balancer (because the `new` hid this detail).